cpp/core/shuffle/Partitioner.cc

```diff
@@ -23,12 +23,13 @@
 
 namespace gluten {
 
-arrow::Result<std::shared_ptr<Partitioner>> Partitioner::make(Partitioning partitioning, int32_t numPartitions) {
+arrow::Result<std::shared_ptr<Partitioner>>
+Partitioner::make(Partitioning partitioning, int32_t numPartitions, int32_t startPartitionId) {
   switch (partitioning) {
     case Partitioning::kHash:
       return std::make_shared<HashPartitioner>(numPartitions);
     case Partitioning::kRoundRobin:
-      return std::make_shared<RoundRobinPartitioner>(numPartitions);
+      return std::make_shared<RoundRobinPartitioner>(numPartitions, startPartitionId);
```

Comment:
Thanks so much for your fix!
As just required by `RoundRobinPartitioner`, can we move the code of generating `startPartitionId` from scala side to native side, i.e., into `RoundRobinPartitioner`? It may be a cleaner way if feasible. Thanks!

Reply:
Thanks for your comments. Yes, I'm firstly also looking to implement this in `RoundRobinPartitioner`, but looks there is no native implement of `XORShiftRandom`. Please point me if there is.
Besides if we move this logic to native side we still need to add a parameter to pass partition id from JVM to native.
Now the native side implment is more like the `HashPartitioner`, and `getPartitionKeyExtractor` are executed in JVM.

Reply:
Yes. There is no native `XORShiftRandom` in Gluten cpp. Since it's just a simple computation of a starting value, we don't need to bother maintaining the same logic on native side. However, the current argument list does look cumbersome. To make it cleaner, we should find a better way to separate parameters for the native `ShuffleWriter` and `Partitioner`, perhaps in a subsequent patch.

Reply:
Thanks for the clarifications. The current change is OK to me.
BTW, I am wondering whether `XORShiftRandom` must be used or can use a replacement (I'm not suggesting to change it).

Reply:
This impacts the partition that rows are placed, I think it's better to use the same algorithm with spark. And it may be required if "SORT_BEFORE_REPARTITION" is true. But there looks another issue, as now we just ignore it, we may add local sort before round robin shuffle in future if `SORT_BEFORE_REPARTITION == true`.
https://github.com/apache/spark/blob/e1a2255f99be88e776295f30f995b339c3e4b5af/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L3367